cmd/transfer/main.go

```
@@ -16,180 +21,294 @@ import (
+		output <- TransferResult{data: lfs.CompleteMessage{
+			Event: "complete",
+			Oid:   msg.Oid,
+			Path:  drsObj.Name,
+		}}
+	}
+}
+
 var Cmd = &cobra.Command{
 	Use:   "transfer",
 	Short: "[RUN VIA GIT LFS] register LFS files into gen3 during git push",
-	Long:  "[RUN VIA GIT LFS] git-lfs transfer mechanism to register LFS files up to gen3 during git push. For new files, creates an indexd record and uploads to the bucket",
+	Long:  `[RUN VIA GIT LFS] git-lfs transfer mechanism to register LFS files up to gen3 during git push. For new files, creates an indexd record and uploads to the bucket`,
 	RunE: func(cmd *cobra.Command, args []string) error {
-		myLogger := drslog.GetLogger()
+		logger := drslog.GetLogger()
+		logger.Print("~~~~~~~~~~~~~ START: drs transfer (SONIC MODE) ~~~~~~~~~~~~~")
 
-		myLogger.Print("~~~~~~~~~~~~~ START: drs transfer ~~~~~~~~~~~~~")
+		numWorkers := getConcurrentTransfers(logger)
 
+		// Gotta go fast — big buffer
 		scanner := bufio.NewScanner(os.Stdin)
-		encoder := json.NewEncoder(os.Stdout)
+		const maxCapacity = 10 * 1024 * 1024 // 10 MB
+		buf := make([]byte, 0, 64*1024)
+		scanner.Buffer(buf, maxCapacity)
+
+		transferQueue := make(chan TransferJob, numWorkers)
+		resultQueue := make(chan TransferResult, numWorkers)
+		var wg sync.WaitGroup
+
+		// Single writer goroutine — must stay ordered
+		writerDone := make(chan struct{})
+		go func() {
+			defer close(writerDone)
+			encoder := encoder.NewStreamEncoder(os.Stdout) // Output still uses stdlib (safe & compatible)
+			for result := range resultQueue {
+				if result.isError {
+					if errMsg, ok := result.data.(lfs.ErrorMessage); ok {
+						lfs.WriteErrorMessage(encoder, errMsg.Oid, errMsg.Error.Code, errMsg.Error.Message)
+					}
+				} else {
+					encoder.Encode(result.data)
+				}
+			}
+		}()
 
+		// Load config
 		cfg, err := config.LoadConfig()
 		if err != nil {
-			myLogger.Printf("Error loading config: %v", err)
+			logger.Printf("Error loading config: %v", err)
```

shouldn't this be down below the first `scanner.Scan()` "init" check? In case it errors then we can return an init error message? The rest above seems fine cause there's no `err`s being returned.